File: crsTransformationCore/src/test/java/com/programmerare/crsTransformations/CrsIdentifierTest.java
package com.programmerare.crsTransformations;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.programmerare.crsTransformations.crsConstants.ConstantCrsCode;
import com.programmerare.crsTransformations.crsConstants.ConstantEpsgNumber;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class CrsIdentifierTest {

    @Test
    void createFromCrsCodeTrimmedButWithoutUppercasingButNotEpsg() {
        CrsIdentifier crsIdentifier = CrsIdentifier.createFromCrsCode("  abc  ");
        assertEquals("abc", crsIdentifier.getCrsCode());
        assertEquals(false, crsIdentifier.isEpsgCode());
    }

    @Test
    void createFromEpsgNumber() {
        // SWEREF99TM = 3006
        CrsIdentifier crsIdentifier = CrsIdentifier.createFromEpsgNumber(3006);
        assertEquals(ConstantCrsCode.SWEREF99TM, crsIdentifier.getCrsCode());
        assertEquals(true, crsIdentifier.isEpsgCode());
        assertEquals(ConstantEpsgNumber.SWEREF99TM, crsIdentifier.getEpsgNumber());
    }

    @Test
    void createFromCrsCodeWithLowerCasedEpsg() {
        CrsIdentifier crsIdentifier = CrsIdentifier.createFromCrsCode("  epsg:4326  ");
        // the input should become trimmed and uppercased "EPSG"
        assertEquals("EPSG:4326", crsIdentifier.getCrsCode());
        assertEquals(true, crsIdentifier.isEpsgCode());
        assertEquals(4326, crsIdentifier.getEpsgNumber());
    }

    @Test
    void createFromCrsCodeNullShouldFail() {
        IllegalArgumentException exception = assertThrows(
            IllegalArgumentException.class,
            () -> CrsIdentifier.createFromCrsCode(null), // should fail
            "Must not be null"
        );
        assertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "non-null");
    }

    @Test
    void createFromCrsCodeWithOnlyWhiteSpaceShouldFail() {
        IllegalArgumentException exception = assertThrows(
            IllegalArgumentException.class,
            () -> CrsIdentifier.createFromCrsCode("   "), // should fail
            "Must not be empty string"
        );
        assertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "non-empty");
    }

    @Test
    void equalsTestWhenCreatingFromEpsgNumberAndFromCrsCode() {
        CrsIdentifier fromEpsgNumber    = CrsIdentifier.createFromEpsgNumber(3006);
        CrsIdentifier fromCrsCode       = CrsIdentifier.createFromCrsCode("  epsg:3006   ");
        assertEquals(fromEpsgNumber, fromCrsCode);
        assertEquals(fromEpsgNumber.hashCode(), fromCrsCode.hashCode());
    }

    /**
     * @param exception
     * @param expectedStringToBeContainedInExceptionMessage e.g. "non-null" or "non-empty"
     */
    private void assertExceptionMessageWhenArgumentWasNullOrEmptyString(
        IllegalArgumentException exception,
        String expectedStringToBeContainedInExceptionMessage
    ) {
        // the exception message is currently something like this: "Parameter specified as non-null is null: method com.programmerare.crsTransformations.CrsIdentifier$Companion.createFromCrsCode, parameter crsCode"
        // (potentially fragile to test the message strings but it does not really change often, and in such a rare scenario, then easy to fix)
        assertThat(exception.getMessage(), containsString(expectedStringToBeContainedInExceptionMessage));
    }
}

File: TestAndExampleProjects/Programmerare.CrsTransformations.Test/CrsTransformations/Identifier/CrsIdentifierTest.cs
namespace Programmerare.CrsTransformations.Identifier {

using NUnit.Framework;
using System;

[TestFixture]
public class CrsIdentifierTest {

    private string GetCrsCodeIncludingUppercasedEpsgPrefix(int epsgNumber) {
        return "EPSG:" + epsgNumber;
    }
    
    [Test]
    public void CrsIdentifier_ShouldReturnWhitespaceTrimmedCrsCodeAndNotBeConsideredAsEpsg_WhenCreatedFromNonEpsgString() {
        CrsIdentifier crsIdentifier = CrsIdentifierFactory.CreateFromCrsCode("  abc  ");
        Assert.AreEqual("abc", crsIdentifier.CrsCode);
        Assert.AreEqual(false, crsIdentifier.IsEpsgCode);
    }

    [Test]
    public void CrsIdentifier_ShouldReturnEpsgNumberAndEpsgPrefixedCrsCodeAndBeConsideredAsEpsg_WhenCreatedFromEpsgNumber() {
        int inputEpsgNumber = 3006;
        // No validation that the number is actually an existing EPSG but any positive integer
        // is assumed to be a EPSG number
        CrsIdentifier crsIdentifier = CrsIdentifierFactory.CreateFromEpsgNumber(inputEpsgNumber);
        Assert.AreEqual(
            inputEpsgNumber, // expected
            crsIdentifier.EpsgNumber
        );
        Assert.AreEqual(GetCrsCodeIncludingUppercasedEpsgPrefix(inputEpsgNumber), crsIdentifier.CrsCode);
        Assert.AreEqual(true, crsIdentifier.IsEpsgCode);
    }

    [Test]
    public void CrsIdentifier_ShouldReturnEpsgNumberAndUppercasedEpsgPrefixedWhitespaceTrimmedCrsCodeAndBeConsideredAsEpsg_WhenCreatedFromLowecasedEpsgCodeWithSurroundingWhitespace() {
        int inputEpsgNumber = 4326;
        string inputCrsCode = "  epsg:" + inputEpsgNumber + "  "; 
        CrsIdentifier crsIdentifier = CrsIdentifierFactory.CreateFromCrsCode(inputCrsCode);
        // the input should become trimmed and return string with uppercased "EPSG:" prefix
        Assert.AreEqual(
            GetCrsCodeIncludingUppercasedEpsgPrefix(inputEpsgNumber), 
            crsIdentifier.CrsCode
        );
        Assert.AreEqual(true, crsIdentifier.IsEpsgCode);
        Assert.AreEqual(inputEpsgNumber, crsIdentifier.EpsgNumber);
    }

    [Test]
    public void CrsIdentifierFactory_ShouldThrowException_WhenCrsCodeInputIsNull() {
        ArgumentException exception = Assert.Throws<ArgumentNullException>(
            () => {
                CrsIdentifierFactory.CreateFromCrsCode(null); // should fail
            }
            , "Must not be null"
        );
        // F# code invoked above may throw exception like this:
        //  nullArg "crsCode"
        // Resulting message: "Value cannot be null. Parameter name: crsCode"
        AssertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "cannot be null");
    }

    [Test]
    public void CrsIdentifierFactory_ShouldThrowException_WhenCrsCodeInputIsOnlyWhitespace() {
        ArgumentException exception = Assert.Throws<ArgumentException>(
            () => {
                CrsIdentifierFactory.CreateFromCrsCode("   "); // should fail
            }
            ,
            "Must not be empty string"
        );
        AssertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "non-empty");
    }

    [Test]
    public void CrsIdentifierFactory_ShouldThrowException_WhenCrsCodeIsEpsgWithNegativeNumber() {
        ArgumentException exception = Assert.Throws<ArgumentException>(
            () => {
                CrsIdentifierFactory.CreateFromCrsCode(GetCrsCodeIncludingUppercasedEpsgPrefix(-123)); // should fail
            }
            ,
            "EPSG must not be negative"
        );
        AssertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "non-positive");
    }


    [Test]
    public void CrsIdentifierFactory_ShouldThrowException_WhenEpsgNumberIsNegative() {
        ArgumentException exception = Assert.Throws<ArgumentException>(
            () => {
                CrsIdentifierFactory.CreateFromEpsgNumber(-1); // should fail
            }
            ,
            "EPSG must not be negative"
        );
        AssertExceptionMessageWhenArgumentWasNullOrEmptyString(exception, "non-positive");
    }

    [Test]
    public void CrsIdentifiers_ShouldBeEqual_WhenCreatedFromEpsgNumberAndCorrespondingCrsCode() {
        CrsIdentifier fromEpsgNumber = CrsIdentifierFactory.CreateFromEpsgNumber(3006);
        CrsIdentifier fromCrsCode = CrsIdentifierFactory.CreateFromCrsCode("  epsg:3006   ");
        Assert.AreEqual(fromEpsgNumber, fromCrsCode);
        Assert.AreEqual(fromEpsgNumber.GetHashCode(), fromCrsCode.GetHashCode());
    }

    private void AssertExceptionMessageWhenArgumentWasNullOrEmptyString(
        ArgumentException exception,
        string expectedStringToBeContainedInExceptionMessage
    ) {
        Assert.NotNull(exception);
        Assert.NotNull(exception.Message);
        // the exception message is currently something like this: "Parameter specified as non-null is null: method Programmerare.CrsTransformations.Identifier.CrsIdentifier$Companion.createFromCrsCode, parameter crsCode"
        // (potentially fragile to test the message strings but it does not really change often, and in such a rare scenario, then easy to fix)
        Assert.That(exception.Message, Does.Contain(expectedStringToBeContainedInExceptionMessage));
        // Alternative to the above test
        StringAssert.Contains(expectedStringToBeContainedInExceptionMessage, exception.Message);
    }

    [Test]
    public void CreateFromCrsCode_ShouldThrowException_WhenCrsCodeIsNull() {
        ArgumentException exception = Assert.Throws<ArgumentNullException>(
            () => {
                CrsIdentifierFactory.CreateFromCrsCode(null);
            }
            ,
            "CRS code must not be null"
        );
    }

    [Test]
    public void CrsIdentifier_ShouldNotBeEqual_WhenDifferentEpsgNumber() {
        Assert.AreNotEqual(
            CrsIdentifierFactory.CreateFromEpsgNumber(123),
            CrsIdentifierFactory.CreateFromEpsgNumber(124)
        );
    }

    [Test]
    public void CrsIdentifier_ShouldNotBeEqual_WhenDifferentCrsCode() {
        Assert.AreNotEqual(
            CrsIdentifierFactory.CreateFromCrsCode("EPSG:987"),
            CrsIdentifierFactory.CreateFromCrsCode("EPSG:986")
        );
    }

    [Test]
    public void crsIdentifier_shouldNotBeEqual_whenDifferentNonEpsgCrsCode() {
        Assert.AreNotEqual(
            CrsIdentifierFactory.CreateFromCrsCode("abc"),
            CrsIdentifierFactory.CreateFromCrsCode("abd")
        );
    }

} // class ends
} // namespace ends